app/src/main/java/fr/free/nrw/commons/settings/SettingsFragment.java

```diff
@@ -141,14 +147,62 @@ private void prepareTheme() {
+     * Uses previously saved language if there is any, if not uses phone locale as initial language.
+     * Adds preference changed listener and saves value chosen by user to shared preferences
+     * to remember later and recall MainActivity to reflect language changes
+     */
+    private void prepareAppLanguages() {
+        final List<String> languageNamesList;
+        final List<String> languageCodesList;
+        final AppLanguageLookUpTable appLanguageLookUpTable = new AppLanguageLookUpTable(
+            Objects.requireNonNull(getContext()));
+        languageNamesList = appLanguageLookUpTable.getLocalizedNames();
+        languageCodesList = appLanguageLookUpTable.getCodes();
+        List<String> languageNameWithCodeList = new ArrayList<>();
+
+        for (int i = 0; i < languageNamesList.size(); i++) {
+            languageNameWithCodeList.add(languageNamesList.get(i) + "[" + languageCodesList.get(i) + "]");
+        }
+
+        final CharSequence[] languageNames = languageNamesList.toArray(new CharSequence[0]);
+        final CharSequence[] languageCodes = languageCodesList.toArray(new CharSequence[0]);
+        // Add all languages and languages codes to lists preference as pair
+        appLangListPreference.setEntries(languageNames);
+        appLangListPreference.setEntryValues(languageCodes);
+
+        // Gets current language code from shared preferences
+        final String languageCode = getCurrentAppLanguageCode();
+        if (languageCode.equals("")){
+            // If current language code is empty, means none selected by user yet so use phone local
+            appLangListPreference.setValue(Locale.getDefault().getLanguage());
+        } else {
+            // If any language is selected by user previously, use it
+            appLangListPreference.setValue(languageCode);
+        }
+
+        appLangListPreference.setOnPreferenceChangeListener((preference, newValue) -> {
+            final String userSelectedValue = (String) newValue;
+            setLocale(Objects.requireNonNull(getActivity()),userSelectedValue);
+            saveAppLanguageValue(userSelectedValue);
+            getActivity().recreate();
+            final Intent intent = new Intent(getActivity(), MainActivity.class);
+            startActivity(intent);
+            return true;
+        });
+    }
+
+    /**
+     * Prepares language summary and language codes list and adds them to list preference as pairs.
+     * Uses previously saved language if there is any, if not uses phone locale as initial language.
      * Adds preference changed listener and saves value chosen by user to shared preferences
      * to remember later
      */
-    private void prepareLanguages() {
-        List<String> languageNamesList;
-        List<String> languageCodesList;
-        AppLanguageLookUpTable appLanguageLookUpTable = new AppLanguageLookUpTable(getContext());
+    private void prepareDescriptionLanguages() {
```

Comment:
Is there no way to factorize both methods? There is a lot of duplicated code.

Reply:
Yeah, I also thought about that and I tried to factorize it but It won't happen that much. But I will try to factorize it once as much as possible.